### PR TITLE
Make the code of compare functions linear in the number of constructors.

### PR DESCRIPTION
**Pull request by aspiwack**
This scheme has been advised by @gashe on #79.

Interestingly there are several comparison functions in Coq which were already implemented with this scheme.

### CODE REVIEW

**Comment by herbelin** [comment]
So, iiuc, not only it makes the code more readable and more regular but also more uniform across the whole system. So all benefit, then?

**Comment by herbelin** [comment]
Maybe pitnicking on esthetical details, but after all, this is precisely a request about style. Somehow I find the version in printer.ml clearer, even if the  lines with same constructors come inbetween, because in the version in globnames.ml, no line with ConstructRef comes in the second block of clauses and it "looks" as if a line about ConstructRef was missing (while the original version in globnames.ml does not give this impression).

**Comment by aspiwack** [comment]
I wouldn't say this is just about style : there is also a concern of tractability. If these functions are to be extended to more constructors, then a function with a quadratic number of cases soon becomes impossible to write.

At any rate, I agree that the interleaved version is clearer. It was just me trying to make minimal commits. I've added a commit to make the `globnames.ml` function more uniform.

**Comment by aspiwack** [comment]
Can this PR be merged?

**Comment by mattam82** [comment]
I'm fine with it, too, I also like the interleaved version better.

File: library/globnames.ml [hunk 1]
```diff
@@ -107,17 +107,16 @@ let global_eq_gen eq_cst eq_ind eq_cons x y =
...
   else match x, y with
+  | VarRef v1, VarRef v2 -> Id.compare v1 v2
+  | VarRef _, _ -> -1
+  | _, VarRef _ -> 1
   | ConstRef cx, ConstRef cy -> ord_cst cx cy
+  | ConstRef _, _ -> -1
+  | _, ConstRef _ -> 1
   | IndRef indx, IndRef indy -> ord_ind indx indy
+  | IndRef _, _ -> -1
+  | _ , IndRef _ -> -1
```
Comment (mattam82) [inline library/globnames.ml]: I'm pretty sure this should be +1 here.

**Comment by mattam82** [comment]
I merge it with the fix

**Pull request closed without merging**